### DISCUSSION TITLE
Is there a way to regex graphemes?

### DISCUSSION THREAD

**Discussion by johnysandels** [karashiiro/TextToTalk#104] [Ideas]
I was wondering if there was a way to use a regex to only replace a grapheme when the regex applies? So I could make the word `hm` get replaced while not messing with the word `hmm` or `Rhythm` . If some sort of Regex is currently possible I could just regex `\s*[hH]m\W` and then replace pronunciation of `hm` only after the regex matches.
The process I was hoping for would be:

```
Grapheme "hm" detected
  if regex `\s*[hH]m\W` matches:
     Replace "hm" with "hm̩"
  else:
    do not replace
```

If this is either already possible, or able to be implemented, it'll be very helpful for short words that are pronounced weirdly, and avoid those short graphemes interacting with longer words that are not included in the lexicon. 
Example, a grapheme for "tis", would interact with satisfy. and would make is say " Sa tis F Y"

https://user-images.githubusercontent.com/79720646/163666419-9a8dfc83-890f-4cf1-a5c1-eef24f830f23.mp4

(May need to unmute)

It would also help with Characters and races with a `'s` contained within their names, that are read normally without needing correcting(thus not in the lexicon) getting broken in the same way with the grapheme `'s`.

I could see possible issues with putting regex into the .pls file if a voice backend isn't expecting those contents and needing a separate file for regexing. If it's already possible though, it may not be a problem, and _may_ already be built into AWS and the system backend. I'm not sure, let me know if you know anything!

**Comment by ryankhart** [comment]
Related discussion here: https://github.com/karashiiro/TextToTalk/discussions/62#discussioncomment-2655298

**Comment by karashiiro** [comment]
Sorry, I missed the notification for this; this would be a deviation from the PLS standard, so it's something that needs to be done carefully. I see the use case, since there are too many permutations of words and punctuation to work within the existing standard on this. On the other hand, allowing regex in our PLS will make our version of PLS incompatible with standard PLS engines, which I feel matters in terms of community stewardship (if someone tries using that lexicon in another program, it will break unexpectedly).

> I could see possible issues with putting regex into the .pls file if a voice backend isn't expecting those contents and needing a separate file for regexing. If it's already possible though, it may not be a problem, and may already be built into AWS and the system backend. I'm not sure, let me know if you know anything!

Given that we already have our own PLS engine that constructs well-formed SSML, supporting this for ourselves shouldn't be a problem.

Some possible solutions:

- Implement an additional XML element for this: in order to maintain compatibility with PLS, we would also [still need](https://www.w3.org/TR/pronunciation-lexicon/#S4.5) a grapheme element in the lexeme. This would be a pain to implement correctly (Should the element be empty? Is that a violation of the requirement that `The <grapheme> element MUST contain 'character' child information items`?)
- Create our own superset of the standard that allows for this: This would be simple enough, so long as we have a different file extension or something to distinguish it. It would also be clear to anyone in the future that this is not PLS-compliant.
- Create a separate standard altogether: We can also just make a [new standard](https://xkcd.com/927/) that's easier to work with and allows for modifications like this.

Converting this into an issue.
